src/command/FileCommands.cxx

```diff
@@ -234,9 +236,93 @@ handle_read_comments(Client &client, Request args, Response &r)
 	gcc_unreachable();
 }
 
+AllocatedPath
+find_art(AllocatedPath directory)
+{
+    std::vector<std::string> art_names = {
+        "cover.png",
+        "cover.jpg",
+        "cover.tiff",
+        "cover.bmp"
+    };
+    
+    for(std::string& name: art_names) {
+        const AllocatedPath art = AllocatedPath::Build(directory, name.c_str());
+        
+        FileInfo fi;
+        if (!GetFileInfo(art, fi, false)) {
+            continue;
+        }
+        
+        if (!fi.IsRegular()) {
+            continue;
+        }
+        return art;
+    }
+    return AllocatedPath::Null();
+}
+
 CommandResult
 handle_album_art(Client &client, Request args, Response &r)
 {
-    r.Error(ACK_ERROR_SYSTEM, "Unimplemented");
-    return CommandResult::ERROR;
+    assert(args.size == 1 || args.size == 2);
+    
+    const char *const uri = args.front();
+    unsigned long long offset = 0;
+    if (args.size == 2) {
+        offset = args.ParseUnsigned(1);
+    }
+    
+#ifdef ENABLE_DATABASE
+    const Storage *storage = client.GetStorage();
+#endif
+    const auto located_uri = LocateUri(uri, &client
+#ifdef ENABLE_DATABASE
+                                       , storage
+#endif
+                                       );
+    
+    const AllocatedPath path = storage->MapFS(located_uri.canonical_uri);
+    const AllocatedPath directory = path.GetDirectoryName();
+    
+    AllocatedPath art_name = find_art(directory);
+    if (art_name.IsNull()) {
+        if (offset > 0) {
+            r.Error(ACK_ERROR_NO_EXIST, "No art found");
+            return CommandResult::ERROR;
+        }
+        r.Write("size: -1\n");
+        return CommandResult::OK;
+    }
+    FileReader *reader = new FileReader(art_name);
+
+    const char *name = PathTraitsFS::GetBase(art_name.c_str());
+    
+    const unsigned long chunk_size = 8192;
```

Comment:
But you didn't use `unsigned long long`, but only `unsigned long`. We know that `unsigned long` is only 32 bit on some platforms (e.g. 32 bit Linux, and all versions of Windows, even 64 bit Windows). We don't know a platform where `unsigned long long` is smaller than 64 bit, but I don't think there's a guarantee on that.
If you want it large-file aware, use `off_t` or `uint64_t` (the latter is what MPD uses internally for file sizes, so let's use that).
It might sound ridiculous to support large files for album art - but that's a matter of "let's do file sizes consistently large-file-aware", a matter of principle. If you break with that rule, it may (and will) haunt you some day in the future.